### PR TITLE
Add keyword args `lt`, `by`, `middle` and `rev` to median

### PR DESCRIPTION
**Pull request by sam0410**
This fixes issue [#19190](https://github.com/JuliaStats/Statistics.jl/issues/26)
A few of the discussions corresponding to this PR are in #19359

### CODE REVIEW

File: stdlib/Statistics/src/Statistics.jl [hunk 1]
```diff
@@ -752,11 +752,11 @@ julia> median([1 2; 3 4], dims=1)
  2.0  3.0
 ```
 """
-median(v::AbstractArray; dims=:) = _median(v, dims)
+median(v::AbstractArray; dims=:, lt=isless, by=identity, middle=middle, rev=false) = _median(v, dims, lt, by, middle, rev)
 
-_median(v::AbstractArray, dims) = mapslices(median!, v, dims = dims)
+_median(v::AbstractArray, dims, lt, by, middle, rev) = mapslices(median!, v, dims = dims)
```
Comment (sam0410) [inline stdlib/Statistics/src/Statistics.jl]: I am not sure on how to handle arguments here. Can someone please help?

File: stdlib/Statistics/src/Statistics.jl [hunk 2]
```diff
@@ -690,33 +690,33 @@ julia> middle(a)
 middle(a::AbstractArray) = ((v1, v2) = extrema(a); middle(v1, v2))
 
 """
-    median!(v)
+    median!(v; lt=isless, by=identity, middle=middle, rev=false)
+
+Compute the median of an array `v`. For an even number of elements no exact
+median element exists, so the result is equivalent to calculating mean of two
```
Comment (StefanKarpinski) [inline stdlib/Statistics/src/Statistics.jl]:
```suggestion
median element exists, so the result is equivalent to calculating the mean of the two
```

File: stdlib/Statistics/src/Statistics.jl [hunk 3]
```diff
@@ -690,33 +690,33 @@ julia> middle(a)
...
+median elements. The `by` keyword lets you provide a function that will be
+applied to each element before comparison. The `lt` keyword allows providing a
+custom "less than" function for sorting prior to taking the median.
+In the case of non-numerical, even length `v`, one can choose either of the two middle
```
Comment (StefanKarpinski) [inline stdlib/Statistics/src/Statistics.jl]:
```suggestion
In the case of non-numerical, even length `v`, the first of the two middle elements
according to the provided ordering is returned as the median. One can choose the
second of the two middle elements by reversing the sort order with `rev=true`.
```

File: stdlib/Statistics/src/Statistics.jl [hunk 4]
```diff
@@ -690,33 +690,33 @@ julia> middle(a)
...
+elements as the median by `rev` keyword.
```
Comment (StefanKarpinski) [inline stdlib/Statistics/src/Statistics.jl]:
```suggestion
```

File: stdlib/Statistics/src/Statistics.jl (hunk 1 above)
Comment (StefanKarpinski) [reply stdlib/Statistics/src/Statistics.jl]: Pass an anonymous function to `mapslices` perhaps?

**Review by StefanKarpinski — CHANGES_REQUESTED** [review]

**Comment by StefanKarpinski** [comment]
Thanks for tackling this—it looks good! I've suggested a few changes to the wording of the docs.

**Comment by sam0410** [comment]
Thanks @StefanKarpinski. I added the changes that you suggested.

**Comment by sam0410** [comment]
I do not think the build is failing because of the code in this PR. It has something to do with pkg/profile.
Thanks for your patience.

**Comment by StefanKarpinski** [comment]
There was a bunch of breakage with the package manager yesterday, I've restarted the CI jobs.

**Comment by pdeffebach** [comment]
Can `lt` be changed to something clearer?

**Comment by StefanKarpinski** [comment]
Sure, but not until 2.0.

**Comment by sam0410** [comment]
`rev` seems more consistent with the existing code. But I can make the change if `tie` seems more appropriate. What do you think @StefanKarpinski ?

**Comment by sam0410** [comment]
Hi, I dropped in to check if this needed more changes and is ready to merge ?

File: stdlib/Statistics/src/Statistics.jl [hunk 5]
```diff
@@ -690,33 +690,34 @@ julia> middle(a)
...
 """
-    median!(v)
+    median!(v; lt=isless, by=identity, middle=middle, rev=false)
+
+Compute the median of an array `v`. For an even number of elements no exact
+median element exists, so the result is equivalent to calculating the mean of the two
+median elements. The `by` keyword lets you provide a function that will be
+applied to each element before comparison. The `lt` keyword allows providing a
+custom "less than" function for sorting prior to taking the median.
+In the case of non-numerical, even length `v`, the first of the two middle elements
```
[the remark this answers is not in the document]
Comment (sam0410) [reply stdlib/Statistics/src/Statistics.jl]: Thanks for the review @nalimilan. I changed the order of description :+1: 
What I meant is that, `rev=true` will change the result only if there is an even length `v` having non numerical values. As in the case of even length `v` having numerical values, the median will be the mean of the 2 middle elements in either case.

**Comment by StefanKarpinski** [comment]
I'd say keep `rev`. We could also add `tie` as an alternative way to express that in the future without breaking anything. Anything else that needs to be done here?

File: stdlib/Statistics/src/Statistics.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (nalimilan) [reply stdlib/Statistics/src/Statistics.jl]: I don't understand what is specific to numerical values here: AFAICT none of the code is dependent on the type of arguments, right?

**Comment by ararslan** [comment]
IMO `rev` is weird in this context. The benefit of a more general `tie` argument is that it could be an actual function, e.g. we could say `tie=first`/`min`, `tie=last`/`max`, or `tie=middle`.

**Comment by StefanKarpinski** [comment]
All the ordering-related APIs take the same keyword arguments, including `rev` so not supporting `rev` is not a viable option. As I said, adding support for some `tie` keyword would also be fine, but that can also be done separately.

File: stdlib/Statistics/src/Statistics.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (sam0410) [reply stdlib/Statistics/src/Statistics.jl]: I was talking about the following. 
```
julia> tt= [1,2,3,4]
4-element Array{Int64,1}:
 1
 2
 3
 4

julia> median(tt)
2.5

julia> median(tt, rev= true)
2.5

julia> median(tt, rev= false)
2.5
```

[the remark this answers is not in the document]
Comment (nalimilan) [reply stdlib/Statistics/src/Statistics.jl]: OK. But then I think the current formulation is confusing, as it should explain that a custom `middle` function has to be provided when `middle` isn't defined for the input type (else you get an error). The second sentence of the paragraph should also make this clear.

AFAICT one can simply pass `middle=first` or `middle=last`, right? Then is `rev` really needed?

**Comment by sam0410** [comment]
I will add the `tie` keyword after this is merged in a seperate PR. Is this PR ready to be merged now?

**Comment by nalimilan** [comment]
My comment [above](https://github.com/JuliaLang/julia/pull/30329#discussion_r245444159) still holds AFAICT.

**Comment by sam0410** [comment]
Thanks for the review @nalimilan . I added the clarity and just like @StefanKarpinski mentioned, I think we must keep `rev`.  Also, I will add `tie` to allow arguments such as `first` and `last` in a later PR.

**Comment by nalimilan** [comment]
Thanks, but I still think some issues haven't been addressed:
- This sentence in the docstring still seems incorrect to me: "In the case of non-numerical, even length `v`,
the first of the two middle elements according to the provided ordering is returned as the median." AFAICT you need to pass `middle=(x, y...) -> x` for that to happen (note the `...` which is needed to accept a single argument when there is an odd number of elements). BTW, an example with non-numerical data would be useful to illustrate how this works.
- I'm still not totally convinced this API is the right one. As the point above shows, using the `middle` argument isn't totally obvious, as simply passing `first` or `last` doesn't work for an odd number of elements. It's even worse if you want to get the last element rather than the first, as one cannot do `middle=(x..., y) -> y`. Maybe better not introduce that `middle` argument if it cannot be used in the typical situation it is designed to handle. (See also previous discussion at https://github.com/JuliaLang/julia/pull/19359.)

**Comment by andreasnoack** [comment]
This would have to be redone against https://github.com/JuliaStats/Statistics.jl

**Pull request closed without merging**